src/audio/drc/drc_generic.c

```diff
@@ -0,0 +1,435 @@
+		abs_input_array[i] = 0.0f;
+		for (ch = 0; ch < nch; ch++) {
+			sample = Q_CONVERT_QTOF(state->pre_delay_buffers[ch][div_start + i], 31);
+			abs_input_array[i] = fmaxf(abs_input_array[i], fabsf(sample));
+		}
+	}
+
+	for (i = 0; i < DRC_DIVISION_FRAMES; i++) {
+		/* Compute compression amount from un-delayed signal */
+		float abs_input = abs_input_array[i];
+
+		/* Calculate shaped power on undelayed input.  Put through
+		 * shaping curve. This is linear up to the threshold, then
+		 * enters a "knee" portion followed by the "ratio" portion. The
+		 * transition from the threshold to the knee is smooth (1st
+		 * derivative matched). The transition from the knee to the
+		 * ratio portion is smooth (1st derivative matched).
+		 */
+		float gain = volume_gain(p, abs_input);
+		int is_release = (gain > detector_average);
+		if (is_release) {
+			if (gain > DRC_NEG_TWO_DB) {
+				detector_average +=
+					(gain - detector_average) *
+					sat_release_rate_at_neg_two_db;
+			} else {
+				float gain_db = linear_to_decibels(gain);
+				float db_per_frame =
+					gain_db * sat_release_frames_inv_neg;
+				float sat_release_rate =
+					decibels_to_linear(db_per_frame) - 1;
+				detector_average += (gain - detector_average) *
+						    sat_release_rate;
+			}
+		} else {
+			detector_average = gain;
+		}
+
+		/* Fix gremlins. */
+		if (isbadf(detector_average))
+			detector_average = 1.0f;
+		else
+			detector_average = min(detector_average, 1.0f);
+	}
+
+	state->detector_average = Q_CONVERT_FLOAT(detector_average, 30);
+}
+
+/* Updates the envelope_rate used for the next division */
+static void drc_update_envelope(struct drc_state *state, struct sof_drc_params *p)
+{
+	const float kA = Q_CONVERT_QTOF(p->kA, 30);
+	const float kB = Q_CONVERT_QTOF(p->kB, 30);
+	const float kC = Q_CONVERT_QTOF(p->kC, 30);
+	const float kD = Q_CONVERT_QTOF(p->kD, 30);
+	const float kE = Q_CONVERT_QTOF(p->kE, 30);
+	const float attack_frames = Q_CONVERT_QTOF(p->attack_frames, 20);
+
+	/* Calculate desired gain */
+	float desired_gain = Q_CONVERT_QTOF(state->detector_average, 30);
+
+	/* Pre-warp so we get desired_gain after sin() warp below. */
+	float scaled_desired_gain = warp_asinf(desired_gain);
+
+	/* Deal with envelopes */
+
+	/* envelope_rate is the rate we slew from current compressor level to
+	 * the desired level.  The exact rate depends on if we're attacking or
+	 * releasing and by how much.
+	 */
+	float envelope_rate;
+
+	int is_releasing = scaled_desired_gain > Q_CONVERT_QTOF(state->compressor_gain, 30);
+
+	/* compression_diff_db is the difference between current compression
+	 * level and the desired level. */
+	float compression_diff_db = linear_to_decibels(
+		Q_CONVERT_QTOF(state->compressor_gain, 30) / scaled_desired_gain);
+
+	if (is_releasing) {
+		/* Release mode - compression_diff_db should be negative dB */
+		state->max_attack_compression_diff_db = INT32_MIN;
+
+		/* Fix gremlins. */
+		if (isbadf(compression_diff_db))
```

Comment:
If you plan to keep floats, would fast-math optimization avoid need for isnormal() check?

Reply:
we are planning to remove floats. just doing floats initially for "get it up and running" then will move to fixed point